### PR TITLE
feat(schedule): replace `later` with `cron`

### LINKED ISSUE #32414: Widen cron syntax support, update docs
### Describe the proposed change(s).

Discussion: https://github.com/renovatebot/renovate/discussions/32369

We should support as much of the cron spec as our parser allows us. I'm not sure why we needed to write our own logic actually, isn't there the ability for us to use a library to tell us "is this now?"

Additionally, our docs on scheduling should be changed to be cron-first and not using `later` syntax by default.

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Replace all later scheduleing in docs & presets with their cron equivalent

## Context
- Closes: https://github.com/renovatebot/renovate/issues/32414

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -2464,7 +2464,7 @@ Here's an example config to limit the "noisy" `aws-sdk` package to weekly update
   "packageRules": [
     {
       "matchPackageNames": ["aws-sdk"],
-      "schedule": ["after 9pm on sunday"]
+      "schedule": ["* 21-24 * * 7*"]
```
Comment (viceice) [inline docs/usage/configuration-options.md:2467]:
```suggestion
      "schedule": ["* 21-24 * * 7"]
```
typo?

File: docs/usage/key-concepts/scheduling.md [hunk 2]
```diff
@@ -93,17 +93,13 @@ Some config examples:
...
 ```
 
 ```json title="Renovate should run outside of common office hours"
 {
-  "schedule": [
-    "after 10pm every weekday",
-    "before 5am every weekday",
-    "every weekend"
-  ]
+  "schedule": ["* 22-23 * * 1-5", "* 0-4 * * 1-5", "* * * * 6,7"]
```
Comment (viceice) [inline docs/usage/key-concepts/scheduling.md:102]:
```suggestion
  "schedule": ["* 22-24 * * 1-5", "* 0-4 * * 0-5", "* * * * 6,7"]
```
is this valid?
```suggestion
  "schedule": ["* 0-5,22-24 * * 1-5", "* * * * 6,7"]
```



do we need `24` or `0` for midnight?

File: docs/usage/noise-reduction.md [hunk 3]
```diff
@@ -107,7 +107,7 @@ Or perhaps at least weekly:
     {
       "matchPackageNames": ["/eslint/"],
       "groupName": "eslint",
-      "schedule": ["before 4am on monday"]
+      "schedule": ["* 0-3 * * 1"]
```
Comment (viceice) [inline docs/usage/noise-reduction.md:110]:
```suggestion
      "schedule": ["* 0-4 * * 1"]
```

File: docs/usage/noise-reduction.md [hunk 4]
```diff
@@ -167,7 +167,7 @@ Let's automerge it if all the linting updates pass:
     {
       "matchPackageNames": ["/eslint/"],
       "groupName": "eslint",
-      "schedule": ["before 4am on monday"],
+      "schedule": ["* 0-3 * * 1"],
```
Comment (viceice) [inline docs/usage/noise-reduction.md:170]:
```suggestion
      "schedule": ["* 0-4 * * 1"],
```

File: lib/config/presets/internal/schedule.ts [hunk 5]
```diff
@@ -2,18 +2,14 @@ import type { Preset } from '../types';
...
-const nonOfficeHours = [
-  'after 10pm every weekday',
-  'before 5am every weekday',
-  'every weekend',
-];
-const quarterly = ['every 3 months on the first day of the month'];
-const weekdays = ['every weekday'];
-const weekends = ['every weekend'];
-const yearly = ['every 12 months on the first day of the month'];
+const daily = ['* 0-3 * * *'];
```
Comment (viceice) [inline lib/config/presets/internal/schedule.ts:5]:
```suggestion
const daily = ['* 0-4 * * *'];
```

File: docs/usage/noise-reduction.md (hunk 3 above)
Comment (RahulGautamSingh) [reply docs/usage/noise-reduction.md:110]: before 4am but if we use range 0-4 then 4.15am will be allowed

File: docs/usage/key-concepts/scheduling.md (hunk 2 above)
Comment (RahulGautamSingh) [reply docs/usage/key-concepts/scheduling.md:102]: 24 is not valid for croner and 22-23 is correct as it covers from 10pm to 11.59pm
And 0-5 covers from midnight to 5.59am

File: docs/usage/configuration-options.md [hunk 6]
```diff
@@ -2464,7 +2464,7 @@ Here's an example config to limit the "noisy" `aws-sdk` package to weekly update
   "packageRules": [
     {
       "matchPackageNames": ["aws-sdk"],
-      "schedule": ["after 9pm on sunday"]
+      "schedule": ["* 21-24 * * 7"]
```
Comment (RahulGautamSingh) [inline docs/usage/configuration-options.md:2467]:
```suggestion
      "schedule": ["* 21-23 * * 7"]
```

File: docs/usage/key-concepts/scheduling.md (hunk 2 above)
Comment (RahulGautamSingh) [inline docs/usage/key-concepts/scheduling.md:102]:
```suggestion
  "schedule": ["* 0-4,22-23 * * 1-5", "* * * * 6,7"]
```

File: docs/usage/configuration-options.md [hunk 7]
```diff
@@ -2464,7 +2464,7 @@ Here's an example config to limit the "noisy" `aws-sdk` package to weekly update
   "packageRules": [
     {
       "matchPackageNames": ["aws-sdk"],
-      "schedule": ["after 9pm on sunday"]
+      "schedule": ["* 21-23 * * 7"]
```
Comment (viceice) [inline docs/usage/configuration-options.md:2467]: @rarkins should be we add a human readable description?

so this is effective 21:00 - 23:59

**Comment by RahulGautamSingh** [comment]
Ranges are inclusive. 
For example, 8-11 for an 'hours' entry specifies execution at hours 8, 9, 10, and 11.

File: lib/config/presets/internal/schedule.ts [hunk 8]
```diff
@@ -2,18 +2,14 @@ import type { Preset } from '../types';
...
+const earlyMondays = ['* 0-3 * * 1'];
+const monthly = ['* 0-3 1 * *'];
+const nonOfficeHours = ['* 0-4,22-23 * * 1-5', '* * * * 6,7'];
+const quarterly = ['* * * */3 *'];
+const weekdays = ['* * * * 1-5'];
+const weekends = ['* * * * 6,7'];
+const yearly = ['* * 1 */12 *'];
```
Comment (viceice) [inline lib/config/presets/internal/schedule.ts:12]: we should add unit tests to ensure the textual representation matches our expectations.

**Comment by RahulGautamSingh** [comment]
- Blocked by: https://github.com/renovatebot/renovate/pull/32755

File: docs/usage/configuration-options.md (hunk 7 above)
Comment (rarkins) [reply docs/usage/configuration-options.md:2467]: Add where?

Comment (viceice) [reply docs/usage/configuration-options.md:2467]: yes

Comment (RahulGautamSingh) [reply docs/usage/configuration-options.md:2467]: We already explain the schedule in the above statement.

File: lib/config/presets/internal/schedule.ts (hunk 8 above)
Comment (viceice) [reply lib/config/presets/internal/schedule.ts:12]: don't see tests yet

**Review by viceice — COMMENTED** [review]
see new comments

File: lib/config/presets/internal/schedule.ts (hunk 8 above)
Comment (RahulGautamSingh) [reply lib/config/presets/internal/schedule.ts:12]: They won't be in this PR...another PR was merged with test file [here](https://[github.com/renovatebot/renovate/pull/32755](https://github.com/renovatebot/renovate/pull/32755))

**Review by viceice — COMMENTED** [review]
please add the mentioned descriptions for better copying as discussed on slack

File: docs/usage/key-concepts/scheduling.md [hunk 9]
```diff
@@ -93,17 +93,15 @@ Some config examples:
...
 
 ```json title="Renovate should run outside of common office hours"
 {
-  "schedule": [
-    "after 10pm every weekday",
-    "before 5am every weekday",
-    "every weekend"
-  ]
+  "description": "Schedule during typical non-office hours on weekdays (i.e., 10 PM - 5 AM) and anytime on weekends",
+  "schedule": ["* 0-4,22-23 * * 1-5", "* * * * 6,7"]
```
Comment (viceice) [inline docs/usage/key-concepts/scheduling.md:104]:
```suggestion
  "schedule": ["* 0-4,22-23 * * 1-5", "* * * * 0,6"]
```
we should consistently use `0,6` or `6,7`

**Review by viceice — COMMENTED** [review]
@rarkins do you prefer zero or seven for sunday? we should be consistent and only use one of them even both are working.

**Comment by HonkingGoose** [comment]
I agree that we should be consistent with the times, if possible. 🙂

File: docs/usage/configuration-options.md [hunk 10]
```diff
@@ -2463,8 +2463,9 @@ Here's an example config to limit the "noisy" `aws-sdk` package to weekly update
 {
   "packageRules": [
     {
+      "description": "Schedule aws-sdk updates on Sunday nights (9 PM - 12 AM)",
       "matchPackageNames": ["aws-sdk"],
-      "schedule": ["after 9pm on sunday"]
+      "schedule": ["* 21-23 * * 7"]
```
Comment (RahulGautamSingh) [inline docs/usage/configuration-options.md:2468]:
```suggestion
      "schedule": ["* 21-23 * * 0"]
```

File: lib/config/presets/internal/schedule.ts [hunk 11]
```diff
@@ -2,76 +2,77 @@ import type { Preset } from '../types';
...
-const weekdays = ['every weekday'];
-const weekends = ['every weekend'];
-const yearly = ['every 12 months on the first day of the month'];
+const daily = ['* 0-3 * * *'];
+const earlyMondays = ['* 0-3 * * 1'];
+const monthly = ['* 0-3 1 * *'];
+const nonOfficeHours = ['* 0-4,22-23 * * 1-5', '* * * * 6,7'];
+const quarterly = ['* * 1 */3 *'];
+const weekdays = ['* * * * 1-5'];
+const weekends = ['* * * * 6,7'];
```
Comment (RahulGautamSingh) [inline lib/config/presets/internal/schedule.ts:11]:
```suggestion
const weekends = ['* * * * 0,6'];
```

File: lib/config/presets/internal/schedule.ts [hunk 12]
```diff
@@ -2,76 +2,77 @@ import type { Preset } from '../types';
...
-  'every weekend',
-];
-const quarterly = ['every 3 months on the first day of the month'];
-const weekdays = ['every weekday'];
-const weekends = ['every weekend'];
-const yearly = ['every 12 months on the first day of the month'];
+const daily = ['* 0-3 * * *'];
+const earlyMondays = ['* 0-3 * * 1'];
+const monthly = ['* 0-3 1 * *'];
+const nonOfficeHours = ['* 0-4,22-23 * * 1-5', '* * * * 6,7'];
```
Comment (RahulGautamSingh) [inline lib/config/presets/internal/schedule.ts:8]:
```suggestion
const nonOfficeHours = ['* 0-4,22-23 * * 1-5', '* * * * 0,6'];
```

**Review by viceice — APPROVED** [review]

**Pull request merged**